### PR TITLE
Tc2 p1 back 

### PR DESCRIPTION
**Pull request by CristobalMe**
## Description
This PR is the backend for the past PR.
It adds the necessary routes to work with "ridesharing"
It also updates the data model for the TC2

## Milestones
TC2

### CODE REVIEW

File: backend/routes/attendanceRoutes.js [hunk 1]
```diff
@@ -141,4 +141,47 @@ router.delete('/user/:userAttending/:eventId', async (req, res) => {
   res.json(attendance)
 })
 
+router.post('/ridesharing/manyIds', async (req, res) => {
```
[the remark this answers is not in the document]
Comment (toriqyanM) [reply backend/routes/attendanceRoutes.js:144]: This should be in ridesharingRoutes. Which route a specific endpoint belongs to is related to what it returns. It is returning ridesharing here.

File: backend/routes/attendanceRoutes.js [hunk 2]
```diff
@@ -141,4 +141,47 @@ router.delete('/user/:userAttending/:eventId', async (req, res) => {
...
+  } catch (error) {
+    console.log(error);
+
+    return res.status(500).json({
+      success: false,
+      message: error.message,
+    });
+  }
+  console.log(ids)
+  console.log(attendance)
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/attendanceRoutes.js:164]: Sorry

File: backend/routes/ridesharingRoutes.js [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+  let ridesharingId = []
+  const { userAttending } = req.params
+
+  // Get all the attendances in wich we have registered preferences
+  try {
+      attendance = await prisma.attendance.findMany({
+        where: { 
+          userAttending: userAttending,
+          ridesharingUserPreferencesForEvent: {isNot: null},
+          ridesharingId: {gt: 0}
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:168]: In none. But prisma required that I added the ridesharingId parameter

File: backend/routes/attendanceRoutes.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/attendanceRoutes.js:164]: Done

File: backend/routes/ridesharingRoutes.js [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+  const arrivalTimeDateTime = new Date(arrivalTime)
+
+  attendance = await prisma.attendance.findMany({
+    where: { 
+      userAttending: user.username,
+      eventId: parseInt(event.id)
+    }
+  })
+  console.log(attendance)
+  console.log(attendance[0].id)
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:18]: Done

File: backend/routes/attendanceRoutes.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/attendanceRoutes.js:144]: To do:
Rework attendanceRoutes.js /ridesharing/manyIds

File: backend/routes/ridesharingRoutes.js [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+const prisma = new PrismaClient()
+
+router.post('/withCar', async (req, res) => {
+  const {  seatsAvailable, arrivalTime, departingLat, departingLong, user, event } = req.body
+  let newRidesharing = []
+  const arrivalTimeDateTime = new Date(arrivalTime)
+
+  attendance = await prisma.attendance.findMany({
+    where: { 
+      userAttending: user.username,
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:13]: To do: 
Solve https://github.com/CristobalMe/Event_Finder/pull/14#discussion_r1671269207

File: backend/routes/ridesharingRoutes.js [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+    attendance.map((a) => {
+        ridesharingId.push(a.id)
+    })
+    try {
+        ridesharing = await prisma.ridesharing.findMany({
+            where: {
+                attendanceId: { in: ridesharingId }
+            }
+        })
+    } catch (error) {
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:145]: To do:
Learn and use include in ridesharingRoutes.js

File: backend/routes/ridesharingRoutes.js [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+      console.log(error);
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:105]: To do:
Learn and use include in ridesharingRoutes.js

File: backend/routes/ridesharingRoutes.js [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+      ridesharingId.push(a.ridesharingId)
+  })
+  // ---------------------------------
+  try {
+      ridesharing = await prisma.ridesharing.findMany({
+          where: {
+            id: { in: ridesharingId }
+          }
+      })
+  } catch (error) {
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:191]: To do:
Learn and use include in ridesharingRoutes.js

File: backend/routes/ridesharingRoutes.js [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+        success: false,
+        message: error.message,
+      });
+  }
+  // -------------------------------------------------------------
+  // Get the ids 
+  attendance.map((a) => {
+      ridesharingId.push(a.ridesharingId)
+  })
+  // ---------------------------------
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:184]: To do:
use for getAttendingRideshares function ridesharingRoutes.js

File: backend/routes/ridesharingRoutes.js [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+  let ridesharingId = []
+  let eventId = []
+  const { userAttending } = req.params
+  let ridesharingAttending = []
+
+  // get ridesharingAttending
+  ridesharingAttending = await getAttendingRideshares(userAttending)
+  // -------------------------
+
+  // recommend ridesharings
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:244]: To do:
Leave more comments in this recommendation logic

File: backend/routes/ridesharingRoutes.js [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+  // -------------------------------------------------------------
+  // Get the ids 
+  attendance.map((a) => {
+      ridesharingId.push(a.ridesharingId)
+  })
+  return(ridesharingId)
+}
+
+// This reccommends posible ridesharings to users without car 
+router.get('/user/notDriving/suggestions/:userAttending', async (req, res) => {
```
[the remark this answers is not in the document]
Comment (CristobalMe) [reply backend/routes/ridesharingRoutes.js:232]: To do:
Re-work recommending ridesharings for a user

**Pull request merged**